test/cpp/tensorexpr/test_reductions.cpp

```diff
@@ -0,0 +1,430 @@
+#include <limits>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <unordered_map>
+#include "test/cpp/tensorexpr/test_base.h"
+
+#include "test/cpp/tensorexpr/padded_buffer.h"
+#include "torch/csrc/jit/tensorexpr/buffer.h"
+#include "torch/csrc/jit/tensorexpr/eval.h"
+#include "torch/csrc/jit/tensorexpr/function.h"
+#include "torch/csrc/jit/tensorexpr/ir.h"
+#include "torch/csrc/jit/tensorexpr/ir_printer.h"
+#include "torch/csrc/jit/tensorexpr/ir_simplifier.h"
+#include "torch/csrc/jit/tensorexpr/loopnest.h"
+#include "torch/csrc/jit/tensorexpr/tensor.h"
+
+namespace torch {
+namespace jit {
+
+using namespace torch::jit::tensorexpr;
+
+// Sum an array to a single value.
+void testReduceSum1D() {
+  KernelScope kernel_scope;
+
+  Buffer b(VarHandle("b", kHandle), kFloat, {10});
+  std::vector<float> in(10);
+  for (int j = 0; j < 10; ++j) {
+    in[j] = j;
+  }
+
+  std::vector<float> out(1, -1.f);
+
+  Tensor* c = ComputeReduce<Add>("sum", {}, 0, b, {{10, "m"}});
```

Comment:
I think it is important that "b" could be a lambda that produces an expression. Although we don't need to deal with Matmul immediately, we would want our frontend to naturally extend to that, since we might get to it in several weeks. That might be the recommended way to use this API.

Also "Add" should imply the initial value of "0". Any custom reducer in the future should imply their own initial values.

Reply:
Yes, there is a lambda version of the `ComputeReduce` call - I just added a new unit test that covers Matmul, seems to map pretty well actually. Looks like this:

```
Buffer tA(VarHandle("tA", kHandle), kFloat, {3, 2});
Buffer tB(VarHandle("tB", kHandle), kFloat, {2, 3});
Tensor* mm = ComputeReduce(
    "mm",
    kFloat,
    {{3, "m"}, {3, "n"}},
    0,
    [&](ExprHandle accum,
        const VarHandle& m,
        const VarHandle& n,
        const VarHandle& k) { return accum + (tA(m, k) * tB(k, n)); },
    {{2, "k"}});
```

Re initial values, it feels a bit more flexible to have them be explicit, but its probably very rare that they're useful. In that case we lose the code size benefits of having the interaction op be a template parameter and it might be better to have ComputeSum, ComputeMax, etc etc - still with a ComputeReduce that takes a lambda.